qupath-core/src/main/java/qupath/lib/analysis/images/ContourTracing.java

```diff
@@ -1254,395 +1277,100 @@ private static boolean selected(double v, double min, double max) {
-				 * or moved for a location 0 to location X - possibly involving merges or completion of 
-				 * an outline.
-				 * 
-				 * Note that outlines are always drawn so that the 'on' pixels are on the left, 
-				 * from the point of view of the directed line.
-				 * Therefore shells are anticlockwise whereas holes are clockwise.
-				 */
-				
-				// Extract the local 2x2 binary pattern
-				// This represented by a value between 0 and 15, where bits indicate if a pixel is selected or not
-				int pattern = (prevRow[left] ? 8 : 0) 
-								+ (prevRow[center] ? 4 : 0) 
-								+ (thisRow[left] ? 2 : 0)
-								+ (thisRow[center] ? 1 : 0);
+		var factory = GeometryTools.getDefaultFactory();
+		var pm = factory.getPrecisionModel();
 
-				
-				switch (pattern) {
-				case 0: 
-					// Nothing selected
-//					assert movingDown[x] == null;
-//					assert movingRight == null;
-					break;
-				case 1: 
-					// Selected D
-//					assert movingDown[x] == null;
-//					assert movingRight == null;
-					// Create new shell
-					movingRight = new Outline(xOffset, yOffset);
-					movingRight.append(x, y);
-					movingDown[x] = movingRight;
-					break;
-				case 2: 
-					// Selected C
-//					assert movingDown[x] == null;
-					movingRight.prepend(x, y);
-					movingDown[x] = movingRight;
-					movingRight = null;
-					break;
-				case 3: 
-					// Selected C, D
-//					assert movingDown[x] == null;
-//					assert movingRight != null;
-					break;
-				case 4: 
-					// Selected B
-//					assert movingRight == null;
-					movingDown[x].append(x, y);
-					movingRight = movingDown[x];
-					movingDown[x] = null;
-					break;
-				case 5: 
-					// Selected B, D
-//					assert movingRight == null;
-//					assert movingDown[x] != null;
-					break;
-				case 6: 
-					// Selected B, C
-//					assert movingDown[x] != null;
-//					assert movingRight != null;
-					movingRight.prepend(x, y);
-					if (Objects.equals(movingRight, movingDown[x])) {
-						// Hole completed!
-						manager.addHole(movingRight);
-						movingRight = new Outline(xOffset, yOffset);
-						movingRight.append(x, y);
-						movingDown[x] = movingRight;
-					} else {
-						movingDown[x].append(x, y);
-						var temp = movingRight;
-						movingRight = movingDown[x];
-						movingDown[x] = temp;
+		int xStart = 0;
+		int yStart = 0;
+		int xEnd = image.getWidth();
+		int yEnd = image.getHeight();
+		// Clip searched pixels using the envelope if provided
+		if (envelope != null) {
+			xStart = Math.max(xStart, (int)Math.floor(envelope.getMinX()-1));
+			yStart = Math.max(yStart, (int)Math.floor(envelope.getMinY()-1));
+			xEnd = Math.min(xEnd, (int)Math.ceil(envelope.getMaxX())+1);
+			yEnd = Math.min(yEnd, (int)Math.ceil(envelope.getMaxY())+1);
+		}
+
+		List<LineString> lines = new ArrayList<>();
+		Coordinate lastHorizontalEdgeCoord = null;
+		Coordinate[] lastVerticalEdgeCoords = new Coordinate[xEnd-xStart+1];
```

Comment:
Is there a reason why the horizontal and vertical coords are treated differently here (one sequence stored, one not)?

Reply:
Yes, we need to know if the last pixel was an edge so that we can continue/end the edge. That's easy when traversing horizontally, but we need to store for an entire row when traversing vertically.

My original implementation looped through all the pixels twice (horizontally then vertically) so then no array was required, but it seemed to be doing unnecessary work - and to therefore be a bit slower.

Reply:
Ah okay, yeah that makes sense